src/terminal/adapter/PageManager.cpp

```diff
@@ -0,0 +1,254 @@
+{
+    _buffer.SetCurrentAttributes(attr);
+    // If the api parameter was specified, we need to pass the new attributes
+    // through to the api. This occurs when there's a potential for the colors
+    // to be changed, which may require some legacy remapping in conhost.
+    if (api)
+    {
+        api->SetTextAttributes(attr);
+    }
+}
+
+til::CoordType Page::Top() const noexcept
+{
+    // If we ever support vertical window panning, the page top won't
+    // necessarily align with the viewport top, so it's best we always
+    // treat them as distinct properties.
+    return _viewport.top;
+}
+
+til::CoordType Page::Bottom() const noexcept
+{
+    // Similarly, the page bottom won't always match the viewport bottom.
+    return _viewport.bottom;
+}
+
+til::CoordType Page::Width() const noexcept
+{
+    // The page width could also one day be different from the buffer width,
+    // so again it's best treated as a distinct property.
+    return _buffer.GetSize().Width();
+}
+
+til::CoordType Page::Height() const noexcept
+{
+    return Bottom() - Top();
+}
+
+til::CoordType Page::BufferHeight() const noexcept
+{
+    return _buffer.GetSize().Height();
+}
+
+til::CoordType Page::XPanOffset() const noexcept
+{
+    return _viewport.left;
+}
+
+til::CoordType Page::YPanOffset() const noexcept
+{
+    return 0; // Vertical panning is not yet supported
+}
+
+PageManager::PageManager(ITerminalApi& api, Renderer& renderer) noexcept :
+    _api{ api },
+    _renderer{ renderer }
+{
+}
+
+void PageManager::Reset()
+{
+    _activePageNumber = 1;
+    _visiblePageNumber = 1;
+    _buffers = {};
+}
+
+Page PageManager::Get(const til::CoordType pageNumber) const
+{
+    const auto requestedPageNumber = std::min(std::max(pageNumber, 1), MAX_PAGES);
+    auto [visibleBuffer, visibleViewport, isMainBuffer] = _api.GetBufferAndViewport();
+
+    // If we're not in the main buffer (either because an app has enabled the
+    // alternate buffer mode, or switched the conhost screen buffer), then VT
+    // paging doesn't apply, so we disregard the requested page number and just
+    // use the visible buffer (with a fixed page number of 1).
+    if (!isMainBuffer)
+    {
+        return { visibleBuffer, visibleViewport, 1 };
+    }
+
+    // If the requested page number happens to be the visible page, then we
+    // can also just use the visible buffer as is.
+    if (requestedPageNumber == _visiblePageNumber)
+    {
+        return { visibleBuffer, visibleViewport, _visiblePageNumber };
+    }
+
+    // Otherwise we're working with a background buffer, so we need to
+    // retrieve that from the buffer array, and resize it to match the
+    // active page size.
+    const auto pageSize = visibleViewport.size();
+    auto& pageBuffer = _getBuffer(requestedPageNumber, pageSize);
+    return { pageBuffer, til::rect{ pageSize }, requestedPageNumber };
+}
+
+Page PageManager::ActivePage() const
+{
+    return Get(_activePageNumber);

```

Comment:
> However, I believe this cannot be done right now since the `visibleViewport` and `isMainBuffer` are managed by the `_api`.

Yeah, I had hoped that one day we might avoid this issue by having `AdaptDispatch` register some sort of callback with the api, so it can get notified of buffer and viewport changes, instead of having to query them every time. But we always have to deal with the fact that the legacy console APIs are in control of those aspects of the architecture, so that information has to be passed across the api boundary one way or the other. Again, not something I want to try and solve now.

However, if it's the copying of these page objects that is causing the performance regression, that might still be something I can improve. When I have more time I'll do some experimenting and see if I can find a way to make that any faster.